src/transform/src/fusion/reduce.rs

```diff
@@ -0,0 +1,105 @@
+// Copyright Materialize, Inc. and contributors. All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+//! Fuses reduce operators with parent operators if possible.
+
+use crate::TransformArgs;
+use expr::{AggregateExpr, AggregateFunc, MirRelationExpr, MirScalarExpr};
+
+/// Fuses reduce operators with parent operators if possible.
+#[derive(Debug)]
+pub struct Reduce;
+
+impl crate::Transform for Reduce {
+    fn transform(
+        &self,
+        relation: &mut MirRelationExpr,
+        _: TransformArgs,
+    ) -> Result<(), crate::TransformError> {
+        relation.visit_mut_pre(&mut |e| {
+            self.action(e);
+        });
+        Ok(())
+    }
+}
+
+impl Reduce {
+    /// Fuses reduce operators with parent operators if possible.
+    pub fn action(&self, relation: &mut MirRelationExpr) {
+        if let MirRelationExpr::Reduce {
+            input,
+            group_key,
+            aggregates,
+            monotonic: _,
+            expected_group_size: _,
+        } = relation
+        {
+            if let MirRelationExpr::Reduce {
+                input: inner_input,
+                group_key: inner_group_key,
+                aggregates: inner_aggregates,
+                monotonic: _,
+                expected_group_size: _,
+            } = &mut **input
+            {
+                // Do nothing if outer keys are not a subset of inner keys
+                if !group_key.iter().all(|k| inner_group_key.contains(k)) {
+                    return;
+                }
+
+                if aggregates.is_empty() && inner_aggregates.is_empty() {
+                    // Drop redundant inner reduce
+                    *input = Box::new(inner_input.take_dangerous());
```

Comment:
To maintain correctness, you need to project the columns. Besides doing the reduction itself, the reduce has an implicit project  `[<group keys in order> <agg columns in order>]`

Below is the original plan for one of the failing sqllogictest tests
```
explain SELECT DISTINCT cor0.col0 FROM tab0  AS cor0 GROUP BY cor0.col1, cor0.col0; 
        Optimized Plan        
------------------------------
%0 =                        +
 | Get blah.public.tab0 (u12)+
 | Distinct group=(#1, #0)   +
 | Distinct group=(#1)       +
```

You could replace the inner distinct with a project
```
explain SELECT DISTINCT cor0.col0 FROM tab0  AS cor0 GROUP BY cor0.col1, cor0.col0; 
        Optimized Plan        
------------------------------
%0 =                        +
 | Get blah.public.tab0 (u12)+
 | Project group=(#1, #0)   +
 | Distinct group=(#1)       +
```

and then projection lifting should change the above to become:
```
explain SELECT DISTINCT cor0.col0 FROM tab0  AS cor0 GROUP BY cor0.col1, cor0.col0; 
        Optimized Plan        
------------------------------
%0 =                        +
 | Get blah.public.tab0 (u12)+
 | Distinct group=(#0)       +
```

Or you could do the permutation yourself.

Reply:
Thanks! I was adding the permutation myself, but I'll also try the other approach.

Reply:
I went with replacing it with a project 👍🏼